### PR TITLE
prov/gni: FI_MORE support for read_msg

### PR DESCRIPTION
**Pull request by a-abraham**
Code to support FI_MORE for reads in gnix_rma.c
Corresponding test created in rdm_dgram_rma.c

@sungeunchoi @ztiffany

### CODE REVIEW

File: prov/gni/src/gnix_rma.c
Comment (sungeunchoi) [inline prov/gni/src/gnix_rma.c:921]: Is it possible to have used all the entries and have `tail_len != 0`?  If so, isn't there a problem in this code?

**Comment by sungeunchoi** [comment]
What about FI_RMA_EVENT?

File: prov/gni/src/gnix_rma.c
Comment (jswaro) [inline prov/gni/src/gnix_rma.c]: At this point, with a function this small, perhaps we could make this an inline function?

File: prov/gni/src/gnix_rma.c
Comment (a-abraham) [reply prov/gni/src/gnix_rma.c:921]: In __gnix_rma_more_fill_pd   there is a loop that calculates the number of entries required. (See 915-952)
+1 for each req.
+1 if head unaligned.
+1 if tail unaligned.  
I don't think there can be a situation where we've run out of entries.

Comment (sungeunchoi) [reply prov/gni/src/gnix_rma.c:921]: ok -- then let's put in an assertion here just to aid in future debugging.

File: prov/gni/src/gnix_rma.c
Comment (ztiffany) [inline prov/gni/src/gnix_rma.c:1101]: Missing ?
+			idx++;
￼

Comment (a-abraham) [reply prov/gni/src/gnix_rma.c:1101]: Fixed.

**Comment by a-abraham** [comment]
Marking as do not merge while I do additional testing and create additional tests

**Comment by a-abraham** [comment]
I've added a second commit to show the additional code I created to handle  Indirect messaging (unaligned get with len <= 60).

I will squash both into one commit before merging.

File: prov/gni/src/gnix_rma.c
Comment (ztiffany) [inline prov/gni/src/gnix_rma.c]: misaligned

File: prov/gni/src/gnix_rma.c
Comment (ztiffany) [inline prov/gni/src/gnix_rma.c]: I suggest:
```
if (!indirect) {
  if (head_off)
    ...
  if(tail_len)
   ...
}
```

File: prov/gni/src/gnix_rma.c
Comment (ztiffany) [inline prov/gni/src/gnix_rma.c]: See refactor suggestion above.

File: prov/gni/test/rdm_dgram_rma.c
Comment (ztiffany) [inline prov/gni/test/rdm_dgram_rma.c]: 'len' isn't used anymore.

**Comment by a-abraham** [comment]
I also plan to remove the GNIX_INFO statements and make them Debug statements where applicable

File: prov/gni/test/rdm_dgram_rma.c
Comment (a-abraham) [reply prov/gni/test/rdm_dgram_rma.c]: fixed

File: prov/gni/src/gnix_rma.c
Comment (a-abraham) [reply prov/gni/src/gnix_rma.c]: Are you saying it should be misaligned instead of aligned?
I think i'll be deleting this statement anyways.

File: prov/gni/src/gnix_rma.c
Comment (a-abraham) [reply prov/gni/src/gnix_rma.c]: fixed

File: prov/gni/src/gnix_rma.c
Comment (a-abraham) [reply prov/gni/src/gnix_rma.c]: fixed

**Comment by hppritcha** [comment]
@a-abraham is there a reason you removed the needs-to-be-merged-upstream label?

File: prov/gni/src/gnix_rma.c
Comment (a-abraham) [reply prov/gni/src/gnix_rma.c:921]: Fixed - added an assertion at the beginning of the code block

File: prov/gni/src/gnix_rma.c
Comment (a-abraham) [reply prov/gni/src/gnix_rma.c]: Sounds good. inline'd

**Comment by a-abraham** [comment]
@hppritcha  I had found a bug I needed to squash, so I removed the tag at that time.
Adding it back now.

File: prov/gni/test/rdm_dgram_rma.c
Comment (sungeunchoi) [inline prov/gni/test/rdm_dgram_rma.c:1684]: Is it possible to replace the constants below with values defined in the provider?  That way, if the defaults change, we are still testing we we think we are testing.

Comment (a-abraham) [reply prov/gni/test/rdm_dgram_rma.c:1684]: messages < GNIX_RMA_UREAD_CHAINED_THRESH     (60) are deemed 'indirect',  and that value is defined in gnix_rma.c

Would I need to import that file in?  Or are other places I should be looking for provider defined values.

Comment (sungeunchoi) [reply prov/gni/test/rdm_dgram_rma.c:1684]: If it's not easy to use the exact values, then can you at least document that that's the value being used and where it is defined.  Then at least a "git grep" will find a reference to it.

Comment (a-abraham) [reply prov/gni/test/rdm_dgram_rma.c:1684]: Okay, I like that idea - I've added the name to the comments.

**Comment by sungeunchoi** [comment]
👍

**Review by ztiffany — APPROVED** [review]

**Pull request merged**
